FEATURE: Allow showing hashtag autocomplete results without term

martin-brennan opened the pull request:
This commit allows us to type # in the UI and present autocomplete
results immediately with the following logic for the topic composer,
and reversed for the chat composer:

* Categories the user can access and has not muted sorted by `topic_count`
* Tags the user can access and has not muted sorted by `topic_count`
* Chat channels the user is a member of sorted by `messages_count`

So in effect, we allow searching for hashtags without a search term.
To do this we add a new `search_without_term` to each data source so
each one can define how it wants to handle this logic.

![chrome-capture-2022-10-29](https://user-images.githubusercontent.com/920448/204451235-3769c455-3cdc-4d40-bf07-e7891d8f5ce7.gif)

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -52,11 +52,43 @@ def self.search(guardian, term, limit)
-        search_results.select { |item| item.slug != term }.sort_by { |item| item.text.downcase },
-      )
+    if term.present?
+      search_results
+        .select { |item| item.slug == term }
+        .sort_by { |item| item.text.downcase }
+        .concat(
+          search_results.select { |item| item.slug != term }.sort_by { |item| item.text.downcase },
+        )
```
```suggestion
      search_results.sort_by do |item|
        [item.slug == term ? 0 : 1, item.text.downcase]
      end
```

I think we can simplify this to a single loop.

martin-brennan replied:
I did try so many variations of this sorting and it still ended up wrong...but I did not try this one so I will see if it will work :+1:

martin-brennan replied:
Ah I see the difference now, I didn't know that you could return multiple values in an array for `sort_by`, this works!

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -52,11 +52,43 @@ def self.search(guardian, term, limit)
+  end
+
+  def self.search_without_term(guardian, limit)
+    guardian_categories = Site.new(guardian).categories
```
`Site#categories` does alot of things which I'm not sure is necessary for the intended use case here. I think we can rely on `User#secure_category_ids` instead.

martin-brennan replied:
The main benefit of `Site.new(guardian).categories` is that all the categories are already cached and we just scope them down to the user, so I don't think there are additional queries involved. I have used it in the other places in `CategoryHashtagDataSource` too. The other benefit is that it already figures out parent category details. But maybe it does make more sense to use that here and in the other places if the query time is negligible?

tgxworld replied:
I think the method is partially cached and it only memoizes the results per instance. Since this is a new instance, we're not going to be benefit from the memoization here. The method does execute extra queries even after fetching all the categories from the class level cache: 

https://github.com/discourse/discourse/blob/e669cf364893421b6545312f62ba7ad5299ee725/app/models/site.rb#L102-L140

martin-brennan replied:
Hmm okay I will rethink this, it's probably fine to just use `User#secure_category_ids` then...thanks for pointing this out!

martin-brennan replied:
I actually would prefer to refactor this in a separate PR too so it's encapsulated as a separate single commit.

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -52,11 +52,43 @@ def self.search(guardian, term, limit)
+    guardian_categories = Site.new(guardian).categories
+
+    category_id_sql = <<~SQL
+      SELECT category_id, MAX(posts.created_at)
+      FROM topics
+      INNER JOIN posts ON posts.topic_id = topics.id
+      WHERE topics.deleted_at IS NULL
+        AND posts.deleted_at IS NULL
+        AND posts.created_at > (NOW() - INTERVAL '2 WEEKS')
+        AND topics.category_id IN (:category_ids)
+      GROUP BY category_id
+      ORDER BY MAX(posts.created_at) DESC
+      LIMIT :limit
```
I'm a litter worried about this query because the performance of the query degrades as the activity of a site increases as we have to join against many posts just to figure out which is the "last active" category. I'm not sure what the requirements are here but are we able to rely on columns like `Category#topic_count`, `Category#post_count` or `Category#lastest_post_id` or `Category#latest_topic_id` to achieve the same result?

martin-brennan replied:
`latest_topic_id` or `latest_post_id` may work here, I will try that out, then we can join the categories on post/topic `created_at` and sort by that, thanks

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -52,11 +52,43 @@ def self.search(guardian, term, limit)
+    guardian_categories
+      .select { |category| category_ids.include?(category[:id]) }
+      .take(limit)
+      .map { |category| category_to_hashtag_item(guardian_categories, category) }
```
Based on the implementation of the method, I'm a little confused about why we need to pass in all the `guardian_categories` instead of just passing the parent category, which we can preload, to the method.

martin-brennan replied:
Ah yeah you are right, will just change it to pass the parent category instead.

tgxworld commented on app/services/tag_hashtag_data_source.rb:
```diff
@@ -55,4 +55,36 @@ def self.search(guardian, term, limit)
+  def self.search_without_term(guardian, limit)
+    return [] if !SiteSetting.tagging_enabled
+
+    tags = DB.query(<<~SQL, limit: limit).map(&:name)
+    SELECT tags.name, MAX(posts.created_at)
+    FROM tags
+    INNER JOIN topic_tags ON topic_tags.tag_id = tags.id
+          INNER JOIN topics ON topic_tags.topic_id = topics.id
+          INNER JOIN posts ON topics.id = posts.topic_id
+          WHERE posts.deleted_at IS NULL AND topics.deleted_at IS NULL
+            AND posts.created_at > (NOW() - INTERVAL '2 WEEKS')
+    GROUP BY tags.name
+    ORDER BY MAX(posts.created_at) DESC
+    LIMIT :limit
+    SQL
```
I have the same concerns as https://github.com/discourse/discourse/pull/19219/files#r1035433392 here but we don't have the same columns we can rely on for the `tags` table. Not sure if we can bring it back to product but I wonder if a simpler heuristic of using `Tag#topic_count` is acceptable.

martin-brennan replied:
Good point, maybe we go by counts instead of recents for this and categories, that will be a lot simpler. Will ask internally.

tgxworld commented on app/services/tag_hashtag_data_source.rb:
```diff
@@ -55,4 +55,36 @@ def self.search(guardian, term, limit)
+        only_tag_names: tags,
+      )
+
+    TagsController
```
This is kind of a code smell here when we're reaching out to a class. Perhaps we should be using a serializer here instead?

martin-brennan replied:
Yeah I really don't like this is in a controller either, I am not really sure why it is. It's used here too:

https://github.com/discourse/discourse/blob/c1ffa4e7b39ad66600440f63c19a6aa7cc2a9f77/app/serializers/detailed_tag_serializer.rb#L8-L10

I will refactor it but I would prefer to do it in a standalone PR

tgxworld commented on plugins/chat/lib/chat_channel_hashtag_data_source.rb:
```diff
@@ -44,4 +44,31 @@ def self.search(guardian, term, limit)
+          exclude_dm_channels: true,
+        )
+      channel_ids = DB.query(<<~SQL, limit: limit).map(&:chat_channel_id)
+        SELECT chat_channel_id, MAX(chat_messages.created_at)
+        FROM chat_messages
+        INNER JOIN chat_channels ON chat_messages.chat_channel_id = chat_channels.id
+        WHERE chat_messages.deleted_at IS NULL
+          AND chat_channels.deleted_at IS NULL
+          AND chat_channel_id IN (#{allowed_channel_ids_sql})
+          AND chat_messages.created_at > (NOW() - INTERVAL '2 WEEKS')
+        GROUP BY chat_channel_id
+        ORDER BY MAX(chat_messages.created_at) DESC
+        LIMIT :limit
```
I think we'll need to store a cache on `chat_channel` with regards to the last created chat message. Otherwise, this is going to be expensive with the number of rows we need to process every single time there is an empty search.

martin-brennan replied:
Ah so a `latest_chat_message_id` on the channel similar to `latest_topic_id` on category?

martin-brennan commented:
@tgxworld I have changed this to use the `topic_count`/`messages_count` columns now and fixed other review comments, there are two outstanding comments but I really would prefer to change those in standalone PRs.

tgxworld commented on app/assets/javascripts/discourse/app/lib/hashtag-autocomplete.js:
```diff
@@ -187,8 +190,10 @@ function _searchGeneric(term, siteSettings, contextualHashtagConfiguration) {
 
-    if (term === "") {
-      return resolve(CANCELLED_STATUS);
+    if (!siteSettings.enable_experimental_hashtag_autocomplete) {
+      if (term === "") {
+        return resolve(CANCELLED_STATUS);
+      }
```
```suggestion
    if (!siteSettings.enable_experimental_hashtag_autocomplete && term === "") {
      return resolve(CANCELLED_STATUS);
```

Minor but removes one level of nesting.

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -48,15 +50,38 @@ def self.search(guardian, term, limit)
+        "LEFT JOIN category_users ON category_users.user_id = #{guardian.user.id}
+        AND category_users.category_id = categories.id",
+      )
+      .where(
+        "NOT categories.read_restricted OR categories.id IN (?)",
+        guardian.user.secure_category_ids,
       )
```
```suggestion
      .secured(guardian)
```

tgxworld commented on spec/services/category_hashtag_data_source_spec.rb:
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+RSpec.describe CategoryHashtagDataSource do
+  fab!(:category1) { Fabricate(:category, slug: "random", topic_count: 432) }
+  fab!(:category2) { Fabricate(:category, slug: "books", topic_count: 235) }
+  fab!(:category3) { Fabricate(:category, slug: "movies", topic_count: 126) }
+  fab!(:group) { Fabricate(:group) }
+  fab!(:category4) { Fabricate(:private_category, slug: "secret", group: group, topic_count: 40) }
+  fab!(:category5) { Fabricate(:category, slug: "casual", topic_count: 99) }
```
I think we can improve the fabrication here by avoiding having `topic_count` be in an increasing order. Otherwise, the ordering can be influenced by the `id` of the topic which follows an increasing order as well.

tgxworld commented on spec/services/category_hashtag_data_source_spec.rb:
```diff
@@ -0,0 +1,36 @@
+
+    it "does not return categories the user does not have permission to view" do
+      expect(described_class.search_without_term(guardian, 5).map(&:slug)).not_to include("secret")
+      GroupUser.create(user: user, group: group)
```
```suggestion
      group.add(user)
```

tgxworld commented on app/services/category_hashtag_data_source.rb:
```diff
@@ -48,15 +50,38 @@ def self.search(guardian, term, limit)
+  end
+
+  def self.search_without_term(guardian, limit)
+    Category
```
I learnt of this while working on https://github.com/discourse/discourse/commit/207b764ea37f13469db1117c5c1102ace05fd2f9 but I'm wondering if we should exclude categories with `Category#search_priority` set to `ignore`.

tgxworld commented on spec/services/category_hashtag_data_source_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    it "does not return categories the user has muted" do
+      CategoryUser.create(
```
```suggestion
      CategoryUser.create!(
```

tgxworld commented on spec/services/hashtag_autocomplete_service_spec.rb:
```diff
@@ -235,6 +236,38 @@ def self.search_sort(search_results, _)
+      fab!(:tag3) { Fabricate(:tag, name: "terrible-books", topic_count: 2) }
+      fab!(:tag4) { Fabricate(:tag, name: "book", topic_count: 1) }
+
+      it "returns the N top categories and tags (based on topic_count) that the user can access" do
```
Just to clarify, what does `N` represent here?

martin-brennan replied:
Changing this, N is irrelevant here, thanks for pointing this out

tgxworld commented on spec/system/hashtag_autocomplete_spec.rb:
```diff
@@ -4,31 +4,52 @@
+
+  it "begins showing results as soon as # is pressed based on categories and tags topic_count" do
+    visit_topic_and_initiate_autocomplete(initiation_text: "#", expected_count: 5)
+    hashtag_results = page.all(".hashtag-autocomplete__link")
```
Since we're using this across multiple tests, perhaps we can move this assertion into a page object?

martin-brennan replied:
Yes will do this when I look at why these system specs are timing out, skipping it for now to fit in with the others

tgxworld commented on app/services/hashtag_autocomplete_service.rb:
```diff
@@ -279,6 +281,23 @@ def lookup_old(slugs)
+      limited_results.concat(set_types(set_refs(search_results), type))
+    end
+
+    limited_results.take(limit)
```
I just noticed we only apply the limit here instead of within the loop. I wonder if it makes sense to terminate the loop early if we already have sufficient results. Otherwise, I think we will always be executing multiple queries each time?

martin-brennan replied:
I only do this here because `split_limit` may end up with an uneven number that is over the acutal limit (e.g. 21 instead of 20) so this just makes sure it meets the cutoff -- individual queries will not return > than the split limits provided (which in the case of `20` is `7` each).

martin-brennan commented:
@tgxworld thanks for approving and the comments, I agree with what you have said here and I will fix up + merge today
